### DISCUSSION TITLE
how to scope preflight and tailwindcss/forms

### DISCUSSION THREAD

**Discussion by samyarkd** [tailwindlabs/tailwindcss#10332] [Help] [answered]
I want to only enable preflight option for some pages and overall i want it to be disabled.
Also i don't need the tailwindcss/forms to be applied global same as preflights i want to limit them for some files

i found a way to limit the preflight, but it wasn't clear enough
i copied the preflight.css into my code but it was't working as expected or it got applied again for the whole app
this is how : https://github.com/tailwindlabs/tailwindcss/discussions/7050#discussioncomment-3140805

please help cuse i researched a lot and i could't find a way to do it.

**Comment by wongjn** [comment, marked as answer]
> i copied the preflight.css into my code but it was't working as expected or it got applied again for the whole app
> this is how : https://github.com/tailwindlabs/tailwindcss/discussions/7050#discussioncomment-3140805

Seems like you did not nest it in a selector scope?

You could also [disable the Preflight core plugin](https://tailwindcss.com/docs/preflight#disabling-preflight), and add your own adapted version of the core plugin:
https://github.com/tailwindlabs/tailwindcss/blob/d0ecd2988637d6bbe4751a6bc60fb579e664da35/src/corePlugins.js#L466-L477
And then scope the selectors when it is parsed into the PostCSS AST and before passing the nodes to `addBase()`.

---

> Also i don't need the tailwindcss/forms to be applied global same as preflights i want to limit them for some files

You could look at using [the class strategy](https://github.com/tailwindlabs/tailwindcss-forms#using-classes-to-style).

**Reply by samyarkd** [reply]
Thank you, yes i did enable nesting and also i wrapped a class around the CSS inside preflight but when i import it into my code 
the whole breaks like the styles are not applying as expected for example heading become small.

> and add your own adapted version of the core plugin

How can i add my own adapted version of core plugins ?

**Reply by wongjn** [reply]
You could use the core plugin code:
https://github.com/tailwindlabs/tailwindcss/blob/d0ecd2988637d6bbe4751a6bc60fb579e664da35/src/corePlugins.js#L466-L477
as a starting point to create your own Tailwind plugin and then scope the selectors when it is parsed into the PostCSS AST and before passing the nodes to `addBase()`.

**Reply by samyarkd** [reply]
Thank you, Finley did it with a bit of help from ChatGPT for the selector part.

This is the solution for others just add this plugin to your plugins and disable the preflight option.:
```cjs
// with class "tailwind-preflight" we can now add preflight only to components we want
// to prevent conflicts between ant design and tailwind

const plugin = require("tailwindcss/plugin");
const fs = require("fs");
const path = require("path");
const postcss = require("postcss");

module.exports = plugin(function ({ addBase }) {
  const preflightStyles = postcss.parse(
    fs.readFileSync(path.join(__dirname, "./preflight.css"), "utf8"),
  );

  // Scope the selectors to specific components
  preflightStyles.walkRules((rule) => {
    rule.selector = ".tailwind-preflight " + rule.selector;
  });

  addBase(preflightStyles.nodes);
});

```
Tailwind docs about plugins: 
https://tailwindcss.com/docs/plugins

**Reply by mohammadhojjat-dev** [reply]
hi,
I got this error when I coped the code to my tailwind css config :

> [vite] Internal server error: [postcss] ENOENT: no such file or directory, open 'F:\Lajward\recharge-app-front-end\recharge-app-front-end\preflight.css'
and where I should put this peace of code in tailwindcss config
?

**Reply by wongjn** [reply]
The error you got is most likely from this line:

```js
fs.readFileSync(path.join(__dirname, "./preflight.css"), "utf8");
```

You may want to replace the `path.join(__dirname, "./preflight.css")` bit if you don't have a `preflight.css` file in the same folder as the file that contains this code. For example, to use the `preflight.css` that comes with the Tailwind package install, you could do `require.resolve('tailwindcss/lib/css/preflight.css')` instead like:

```js
fs.readFileSync(require.resolve('tailwindcss/lib/css/preflight.css'), "utf8");
```

**Reply by gfpacheco** [reply]
Great answer, but I had to update to make it work with rules with multiple selectors:
```javascript
plugin(({ addBase }) => {
  const preflightStyles = postcss.parse(
    fs.readFileSync(require.resolve('tailwindcss/lib/css/preflight.css'), 'utf8'),
  );

  // Scope the selectors to my app
  preflightStyles.walkRules((rule) => {
    rule.selector = rule.selector
      .split(',')
      .map((selector) => `#my-app-root ${selector}`)
      .join(',');
  });

  addBase(preflightStyles.nodes);
})
```

**Comment by Roman86** [comment]
I applied that plugin and got the font size incorrect (at least).

How to resolve the issue with CSS selector specificity?

```css
.tailwind-preflight h1, h2, h3, h4, h5, h6 {
  font-size: inherit;
  font-weight: inherit;
}
```

is more specific than
```css
.text-4xl {
  font-size: 2.25rem;
  line-height: 2.5rem;
}
```

So the font size is incorrect in result.
Probably all the tailwindcss selectors should be prepended with that custom selector somehow?

**Reply by wongjn** [reply]
You could modify the specificity of the `.tailwind-preflight h1` rule, such as by using [`:where()`](https://developer.mozilla.org/en-US/docs/Web/CSS/:where):

```css
:where(.tailwind-preflight h1), h2, h3, h4, h5, h6 {
  /* … */
}

/* Or */

:where(.tailwind-preflight) h1, h2, h3, h4, h5, h6 {
  /* … */
}

/* Or */

.tailwind-preflight :where(h1), h2, h3, h4, h5, h6 {
  /* … */
}

/* Or */

:where(.tailwind-preflight h1, h2, h3, h4, h5, h6) {
  /* … */
}

/* Et cetera */
```

or by using a CSS-native `@layer` (not to be confused with Tailwind's `@layer`):

```css
@layer {
  .tailwind-preflight h1, h2, h3, h4, h5, h6 {
    /* … */
  }
}
```

Otherwise, you could indeed have all the Tailwind CSS-generated rule selectors be prepended with that custom selector.

**Reply by Roman86** [reply]
Thanks for the answer! But I realized that the original solution has a bug (as well as yours). See my [comment](https://github.com/tailwindlabs/tailwindcss/discussions/10332#discussioncomment-6981274)

**Reply by wongjn** [reply]
Yes, I saw noticed that bug too when I first saw the code.

**Comment by Roman86** [comment]
the following code does not do the job correctly:
```js
rule.selector = ".tailwind-preflight " + rule.selector;
```

It should modify **every** selector in the matching rule (there are multiple, comma-separated), not just the first one.
Here is the fixed version:
```js
const plugin = require('tailwindcss/plugin');
const fs = require('fs');
const postcss = require('postcss');

export default function tailwindCssManualPreflightPlugin(preflightSelector) {
  if (!preflightSelector) {
    throw new Error('Selector to manually enable the tailwindcss preflight is not provided');
  }
  const preflightCssPath = require.resolve('tailwindcss/lib/css/preflight.css');
  return plugin(function ({ addBase }) {
    const preflightStyles = postcss.parse(fs.readFileSync(preflightCssPath, 'utf8'));

    // Scope the selectors to specific components
    preflightStyles.walkRules((rule) => {
      rule.selectors = rule.selectors.map((s) => `${s}:where(${preflightSelector} *)`);
      rule.selector = rule.selectors.join(',\n');
    });

    addBase(preflightStyles.nodes);
  });
}

```

Even more enhanced version is here:
https://www.npmjs.com/package/tailwindcss-scoped-preflight

it has 2 modes for CSS selector - "matched only" and "except matched", see Readme

**Reply by OluAgunloye** [reply]
This doesn't seem to work w/ Vite? EMS/CJS errors

**Reply by wongjn** [reply]
It does seem to intermix ESM/CJS syntax. Perhaps convert it to fully one or the other.

**Reply by OluAgunloye** [reply]
How would one do that though? I mean the npm package itself doesn't work on Vite!

**Reply by wongjn** [reply]
If one would want to use ESM syntax, replace the `require()` statements. If one would want to use CJS, replace the `export default`.

**Reply by OluAgunloye** [reply]
The plugin itself uses ESM so when vite attempts to walk the dependency tree the same issue occurs.  See: https://github.com/Roman86/tailwindcss-scoped-preflight/issues/2

You'd have to do this for _all_ of the deps in the tree.

EDIT: You can just change the targets for your Vite imports and exports, but they recommend ESM only. `.mjs` only that means.

**Reply by ryan-saffer** [reply]
this plugin saved me 🙏 i have started using tailwind in an existing large project, and it broke everything. this plugin allows me to progressively add tailwind to new screens only!

**Reply by mryechkin** [reply]
> Even more enhanced version is here:
https://www.npmjs.com/package/tailwindcss-scoped-preflight

Thank you for this! Worked perfectly for me. It's a life saver indeed 😄

**Reply by gagangoku** [reply]
https://www.npmjs.com/package/tailwindcss-scoped-preflight is pretty easy to use, thanks !

**Reply by Roman86** [reply]
small announcement: please try v3 beta version (look for it [on the versions page](https://www.npmjs.com/package/tailwindcss-scoped-preflight?activeTab=versions), click on the last chosen version for the docs). V3 accepts the isolator as a customizable function (there are built-in handlers exported for you to do the same old job). Also it allows to skip particular preflight styles and values if they cause any problems on your side. There is also a migration guide - it's trivial

**Reply by ICEDLEE337** [reply]
this is a great plugin... thank you!

**Reply by jcsuzanne** [reply]
Awesome plugin! Thanks for creating this!

**Reply by donathanqueiros** [reply]
your plugin should be referenced in the documentation.

**Reply by terra-alliance** [reply]
Thank you for the plugin!!!

**Reply by Roman86** [reply]
tailwindcss v4 is now supported (css first, with @plugin with @config support if needed), please check it out, make sure to read the migration guide:
https://www.npmjs.com/package/tailwindcss-scoped-preflight

**Comment by sydneyvaughn75-ops** [comment]
Thanks for reporting this
